src/server/index.js

```diff
@@ -1,188 +1,228 @@
-const { JoinInit, JoinChallenge, JoinChallengeSolution, JoinVerify, Discovery, DiscoveryAck, DialRequest, DialResponse, Error } = require('../rpc/proto')
-
-const prom = (f) => new Promise((resolve, reject) => f((err, res) => err ? reject(err) : resolve(res)))
-
-const sha5 = (data) => crypto.createHash('sha512').update(data).digest()
-
-const crypto = require('crypto')
-const ID = require('peer-id')
-
 const debug = require('debug')
 const log = debug('libp2p:stardust:server')
 
-const handleDial = async (conn, id, server) => {
-  const stream = handshake()
-  pull(
-    conn,
-    stream,
-    conn
-  )
-
-  log('incomming dial')
-
-  const shake = stream.handshake
-  const rpc = LP.wrap(shake, LP.writeWrap(shake.write))
-
-  const { target } = await rpc.readProto(DialRequest)
-  const targetB58 = new ID(target).toB58String()
-
-  log('dial from %s to %s', id.toB58String(), targetB58)
+const Libp2p = require('libp2p')
+const Transport = require('libp2p-websockets')
+const Muxer = require('libp2p-mplex')
+const Secio = require('libp2p-secio')
 
-  const targetPeer = server.network[targetB58]
+const crypto = require('crypto')
+const delay = require('delay')
 
-  if (!targetPeer) {
-    return rpc.writeProto(DialResponse, { error: Error.E_TARGET_UNREACHABLE })
-  }
+const Wrap = require('it-pb-rpc')
+const pipe = require('it-pipe')
+const { int32BEDecode, int32BEEncode } = require('it-length-prefixed')
+const { JoinInit, JoinChallenge, JoinChallengeSolution, JoinVerify, Discovery, DiscoveryAck, DialRequest, DialResponse, Error } = require('../proto')
 
-  try {
-    const conn = await newMuxConn(targetPeer.muxed)
-    rpc.writeProto(DialResponse, {})
-
-    pull(conn, shake.rest(), conn)
-  } catch (e) {
-    return rpc.writeProto(DialResponse, { error: Error.E_GENERIC })
-  }
-}
+const multiaddr = require('multiaddr')
+const PeerId = require('peer-id')
 
-const newMuxConn = (muxed) => prom(cb => muxed.newStream(cb))
-const sleep = (i) => new Promise((resolve, reject) => setTimeout(resolve, i))
+const sha5 = (data) => crypto.createHash('sha512').update(data).digest()
 
-const checkAckLoop = async (rpc, onEnd) => {
+const checkAckLoop = async (wrappedStream, onEnd) => {
   try {
     while (true) {
-      await rpc.readProto(DiscoveryAck)
-      await sleep(100)
+      await wrappedStream.readPB(DiscoveryAck)
+      await delay(100)
     }
   } catch (err) { // we'll get here after a disconnect
     onEnd()
   }
 }
 
+/**
+* Stardust Transport Server
+* @class
+*/
 class Server {
-  constructor ({ transports, addresses, muxers }) {
-    this.switch = new MicroSwitch({ transports, addresses, muxers, handler: this.handler.bind(this) })
-
+  /**
+   * @constructor
+   * @param {Object} options - Options for the listener
+   * @param {Array<multiaddr>} options.addresses
+   * @param {Array<Transport>} options.transports
+   * @param {Array<Multiplexer>} options.muxers
+   * @param {Array<Encryption>} options.encryption
```

Comment:
The idea was to use stardust over wss:// protocol, making it encrypted already, and dropping the extra secio layer.
Connections between peers are always encrypted end-to-end encrypted and authed by secio, so even plaintext stardust isn't posing a security risk (aside from allowing praying eyes to view metadata)

Since it was planned to be used over wss:// the SECIO layer will only slow things down.

Reply:
@mkg20001 yeah, I know that not using the extra encryption layer was one the ideas here. I focused on getting this to work with the new libp2p api first. Now that things are getting shaped, I will understand if we will be able to not use the SECIO layer

Reply:
Ok

Possibly it could just use a switch with plaintext, otherwise making a new micro-switch would also be possible should the libp2p switch be either too complicated to run without a swarm or cause significant performance impacts

We should also look out to prevent the user from having to enter an id, even if it's just a placeholder for plaintext-crypto. If not possible to remove from the switch a fake id could always be appended by the listener in that case.

Reply:
We should make the libp2p dialer working for all types of use cases, so I think that we should not go on creating a new "switch" to get this and understand what can be done without compromising security. In this way, we cannot use the Plaintext crypto protocol as peers would have to add it to use `Stardust` 🙅‍♂

We can potentially have an [upgrader](https://github.com/libp2p/js-libp2p/blob/master/src/upgrader.js) here, that would basically allow us to upgrade the connection manually instead of using the provided upgrader (this would an improvement over the `micro-switch` idea).

I was discussing this with @jacobheun earlier today and he pointed out an important aspect. Our `websocket` transport is relying on the underlying ws module to verify if the data is encrypted and it is not considering `wss` / `ws`. This creates a risk for this.

We should probably think a bit about the alternatives that we have and understand what will be the best decision.

Reply:
> We can potentially have an upgrader here, 

Sounds good

> ws module to verify if the data is encrypted and it is not considering wss / ws. This creates a risk for this.

As I already mentioned earlier the actual p2p connections are encrypted and authed by SECIO

What is possible with using a non-encrypted stardust connection is just DOSing the client with false discovery with a MITM information (this would only work if a MITM would happen and additionally would create at most a minor inconvenience) and the leakage of some metadata (who talks to whom) which would be leaked with regular TCP anyways as the SECIO-handshake is unencrypted.

Stardust is also mainly intended for websites that can't directly dial to peers so the website is most likely using HTTPS which means that we're not _able_ to use unencrypted ws

So unless I'm missing something it appears to be a wrong assumption that this creates a problem.

Though checking if the underlying connection is ws or wss isn't hard as the transport is the one that's dialing so it should be enough to match it the multiaddr in question uses ws or wss

Reply:
So if we insist on encryption we could do:
 - if ws: custom upgrader with SECIO (or user-specified) as only crypto
 - if wss: custom upgrader with plaintext as only crypto

The server would use both plaintext and SECIO then

Reply:
One thing we need to keep in mind is that stardust allows for any `reliable` address, which includes tcp, webrtc, http, etc. We are also not even checking that, just that the stardust multiaddr codec exists in the address. If we remove the potential additional encryption layer, we should be more strict about the addresses we allow. As @mkg20001 mentioned, the general use case is websockets in browser, which will mandate wss. However, I don't see anything in the dialer/listener preventing me from connecting to an insecure tcp stardust server if I discover the address, but as already mentioned it's a low threat attack vector.

We could enforce wss and use plaintext, but the fact that stardust potentially supports any reliable multiaddr is pretty nice. If we want to support broader use cases like `tcp > stardust > websockets` broadly using secio to reduce the complexity might be worth the throughput hit. If we just care about `wss > stardust > websockets`, I think we can enforce that and reasonably use a plaintext upgrader.

Reply:
I suggest .that we take this out of the scope of the refactor PR and discuss this on an issue. I think we should be able to provide options to the transport/server that would allow the users to operate with both of those cases. What do you think? Can you create an issue @mkg20001 ?

Reply:
I think it would be reasonable to take the double encryption hit for now and do a followup release for performance tuning for transports with built in encryption. It's not ideal but mitigates any risk/complexity of multiple reliable multiaddrs and gets the transport usable for folks.